fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,673 @@
+    Returns whether the following condition is true using only integer arithmetic
+    support / maximum_support >
+    0.5 * (1 + proposer_score / maximum_support) + confirmation_byzantine_threshold / 100
+    """
+
+    return (
+        100 * support > 
+        50 * maximum_support + 50 * proposer_score + confirmation_byzantine_threshold * maximum_support 
+    )
+```
+
+#### `is_lmd_confirmed`
+
+```python
+def is_lmd_confirmed(store: Store, confirmation_byzantine_threshold: int, block_root: Root) -> bool:
+    if block_root == store.finalized_checkpoint.root:
+        return True
+    else:
+        block = store.blocks[block_root]
+        finalized_block = store.blocks[store.finalized_checkpoint.root]
+        if block.slot <= finalized_block.slot:
+            # This block is not in the finalized chain.
+            return False
+        else:
+            # Check is_one_confirmed for this block and is_lmd_confirmed for the preceding chain.
+            return (
+                is_one_confirmed(store, confirmation_byzantine_threshold, block_root) and
+                is_lmd_confirmed(store, confirmation_byzantine_threshold, block.parent_root)
+            )
+```
+
+#### `get_total_active_balance_for_block_root`
+
+```python
+def get_total_active_balance_for_block_root(store: Store, block_root: Root) -> Gwei:
+    assert block_root in store.block_states
+
+    # Alternatively, fetch the state from the block's checkpoint
+    state = store.block_states[block_root]
+
+    return get_total_active_balance(state)
+```
+
+#### `get_remaining_weight_in_current_epoch`
+
+```python
+def get_remaining_weight_in_current_epoch(store: Store, block_root: Root) -> Gwei:
+    """ Returns the total weight of votes for this epoch from future committees after the current slot
+    """
+    assert block_root in store.block_states
+
+    # Alternatively, fetches the state from the block's checkpoint
+    state = store.block_states[block_root]
+
+    current_slot = get_current_slot(store)
+    first_slot_next_epoch = compute_start_slot_at_epoch(Epoch(compute_epoch_at_slot(current_slot) + 1))
+    return get_committee_weight_between_slots(state, Slot(current_slot + 1), Slot(first_slot_next_epoch - 1))
+```
+
+#### `get_leaf_block_roots`
+
+```python
+def get_leaf_block_roots(store: Store, block_root: Root) -> Set[Root]:
+    children = [
+        root for root in store.blocks.keys()
+        if store.blocks[root].parent_root == block_root
+    ]
+
+    if any(children):
+        # Get leaves of all children and add to the set.
+        leaves = set().union(*[get_leaf_block_roots(store, child) for child in children])
+        return leaves
+    else:
+        # This block is a leaf.
+        return set([block_root])
+
+```
+
+#### `get_current_epoch_participating_indices`
+
+```python
+def get_epoch_participating_indices(
+    state: BeaconState, 
+    active_validator_indices: Sequence[ValidatorIndex],
+    current_epoch: bool
+) -> Set[ValidatorIndex]: 
```

Comment:
Spec generally uses following format for long function defs:
```suggestion
def get_epoch_participating_indices(state: BeaconState, 
                                    active_validator_indices: Sequence[ValidatorIndex],
                                    current_epoch: bool) -> Set[ValidatorIndex]: 
```

Reply:
I would suggest the function take `epoch` instead of `current_epoch` as a boolean. And assert `epoch in get_current_epoch(state), get_previous_epoch(state)`. I think this makes the function safer and cleaner for the caller

Reply:
I think that this would not be the same as currently `current_epoch` indicates whether the block epoch is the current or previous epoch according to the `store`, not according to the block's state.